planner/core/push_down_predicate_to_table_scan.go

```diff
@@ -0,0 +1,305 @@
+			}
+		}
+	case *PhysicalTableReader:
+		predicatePushDownToTableScan(sctx, p.tablePlan)
+	default:
+		for _, child := range plan.Children() {
+			predicatePushDownToTableScan(sctx, child)
+		}
+	}
+}
+
+/**
+ * @brief: This function is used to remove the empty selection in plan
+ * 		   when late materialization is enabled and all conditions are pushed down.
+ * @param: plan: the physical plan
+ */
+func removeEmptySelection(plan PhysicalPlan) {
+	if tableReader, isTableReader := plan.(*PhysicalTableReader); isTableReader && tableReader.StoreType == kv.TiFlash {
+		if selection, isSelection := tableReader.tablePlan.(*PhysicalSelection); isSelection && len(selection.Conditions) == 0 {
+			if len(tableReader.tablePlan.Children()) == 1 {
+				tableReader.tablePlan = tableReader.tablePlan.Children()[0]
+			} else {
+				logutil.BgLogger().Warn("invalid table plan", zap.String("plan", plan.ExplainID().String()))
+			}
+		}
+		removeEmptySelection(tableReader.tablePlan)
+		tableReader.TablePlans = flattenPushDownPlan(tableReader.tablePlan)
+	} else {
+		var newChildren []PhysicalPlan
+		removed := false
+		for _, child := range plan.Children() {
+			switch x := child.(type) {
+			case *PhysicalSelection:
+				if len(x.Conditions) == 0 {
+					newChildren = append(newChildren, x.children...)
+					removed = true
+				} else {
+					newChildren = append(newChildren, x)
+				}
+			default:
+				newChildren = append(newChildren, x)
+			}
+		}
+		if removed && len(newChildren) > 0 {
+			plan.SetChildren(newChildren...)
+		}
+		for _, child := range plan.Children() {
+			removeEmptySelection(child)
+		}
+	}
+}
+
+/**
+ * @brief: This function is used to transform the columns to a string of "0" and "1".
+ * @param: cols: the columns of a Expression
+ * @param: totalColumnCount: the total number of columns in the tablescan
+ * @example: the columns of tablescan are [a, b, c, d, e, f, g, h]
+ * 		 	 the expression are ["a > 1", "c > 1", "e > 1 or g > 1"]
+ *           so the columns of the expression are [a, c, e, g] and the totalColumnCount is 8
+ *           the return value is "10101010"
+ * @return: the string of "0" and "1"
+ */
+func transformColumnsToCode(cols []*expression.Column, totalColumnCount int) string {
+	code := make([]byte, totalColumnCount)
+	for _, col := range cols {
+		code[col.Index] = '1'
+	}
+	return string(code)
+}
+
+/**
+ *
+ * @brief: This function is used to group the conditions by the column they use
+ *         and sort the groups by the selectivity of the conditions in the group.
+ * @param: conds: the conditions to be grouped
+ * @return: the groups of conditions sorted by the selectivity of the conditions in the group
+ * @example: conds = [a > 1, b > 1, a > 2, c > 1, a > 3, b > 2]
+ *           return = [[a > 3, a > 2, a > 1], [b > 2, b > 1], [c > 1]]
+ * @note: when the selectivity of one group is larger than the threshold,
+ *  	  we will remove it from the returned result.
+ */
```

Comment:
In the TiDB repo, the comment of function header is usually
```
// FunName does ....
// others things.
```

Reply:
revised